### PR TITLE
Add p2p unicast

### PR DESCRIPTION
**Pull request by jean-roland**
Reworked pico internals to accommodate p2p unicast.

You can connect to a list of unicast sockets (eg UDP or TCP) as a peer and listen on a unicast socket (TCP and unix only for now) to accept incoming peer connections.

### CODE REVIEW

File: include/zenoh-pico/system/common/platform.h [hunk 1]
```diff
@@ -69,8 +69,21 @@ void _z_report_system_error(int errcode);
...
+        if (__res != 0) {                  \
+            _z_report_system_error(__res); \
+            return _Z_ERR_SYSTEM_GENERIC;  \
+        }                                  \
+    } while (false)
+
+z_result_t _z_socket_set_non_blocking(const _z_sys_net_socket_t *sock);
+z_result_t _z_socket_accept(const _z_sys_net_socket_t *sock_in, _z_sys_net_socket_t *sock_out);
+void _z_socket_close(_z_sys_net_socket_t *sock);
+z_result_t _z_socket_wait_event(void *peers, void *mutex);
```
Comment (sashacmc) [inline include/zenoh-pico/system/common/platform.h:84]: What the reason to pass parameters as pointers to void instead of pointer to _z_transport_unicast_peer_list_t and _z_mutex_rec_t?

**Review by sashacmc — CHANGES_REQUESTED** [review]
And also we need some example or documentation how to use this p2p unicast mode. It looks like current examples don't allow use it.

File: tests/z_test_unicast.c [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "zenoh-pico.h"
+
+#if Z_FEATURE_SUBSCRIPTION == 1 && Z_FEATURE_PUBLICATION == 1 && Z_FEATURE_QUERY == 1 && Z_FEATURE_QUERYABLE == 1 && \
+    Z_FEATURE_MULTI_THREAD == 1 && Z_FEATURE_LOCAL_SUBSCRIBER == 0 && Z_FEATURE_UNICAST_PEER == 1 &&                 \
+    defined Z_FEATURE_UNSTABLE_API
```
[the remark this answers is not in the document]
Comment (jean-roland) [reply tests/z_test_unicast.c:24]: But it requires the querier

File: src/api/api.c [hunk 3]
```diff
@@ -1037,14 +1044,17 @@ z_result_t z_declare_publisher(const z_loaned_session_t *zs, z_owned_publisher_t
     // Set publisher
     _z_publisher_t int_pub = _z_declare_publisher(zs, key, opt.encoding == NULL ? NULL : &opt.encoding->_this._val,
                                                   opt.congestion_control, opt.priority, opt.is_express, reliability);
-    // Create write filter
-    z_result_t res =
-        _z_write_filter_create(_Z_RC_IN_VAL(zs), &int_pub._filter, keyexpr_aliased, _Z_INTEREST_FLAG_SUBSCRIBERS);
-    if (res != _Z_RES_OK) {
-        if (key._id != Z_RESOURCE_ID_NONE) {
-            _z_undeclare_resource(_Z_RC_IN_VAL(zs), key._id);
+    // TODO: Rework write filters to work with non-aggregated interests
```
[the remark this answers is not in the document]
Comment (jean-roland) [reply src/api/api.c:1047]: Issue is opened here already: #909

File: include/zenoh-pico/system/common/platform.h (hunk 1 above)
Comment (jean-roland) [reply include/zenoh-pico/system/common/platform.h:84]: Sadly it creates a cross include dependency hell.

File: src/collections/list.c [hunk 4]
```diff
@@ -90,6 +91,18 @@ _z_list_t *_z_list_find(const _z_list_t *xs, z_element_eq_f c_f, const void *e)
     return ret;
 }
 
+_z_list_t *_z_list_drop_element(_z_list_t *list, _z_list_t *elem, _z_list_t *prev, z_element_free_f f_f) {
```
[the remark this answers is not in the document]
Comment (jean-roland) [reply src/collections/list.c:94]: The method is safe it's just a way to remove in-place one element of a single linked list.

File: tests/z_test_unicast.c [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    zp_config_insert(z_loan_mut(node_ctx_0.config), Z_CONFIG_MODE_KEY, "peer");
+    zp_config_insert(z_loan_mut(node_ctx_0.config), Z_CONFIG_LISTEN_KEY, "tcp/127.0.0.1:7447");
+    zp_config_insert(z_loan_mut(node_ctx_1.config), Z_CONFIG_MODE_KEY, "peer");
+    zp_config_insert(z_loan_mut(node_ctx_1.config), Z_CONFIG_LISTEN_KEY, "tcp/127.0.0.1:7448");
+    zp_config_insert(z_loan_mut(node_ctx_1.config), Z_CONFIG_CONNECT_KEY, "tcp/127.0.0.1:7447");
+    zp_config_insert(z_loan_mut(node_ctx_2.config), Z_CONFIG_MODE_KEY, "peer");
+    zp_config_insert(z_loan_mut(node_ctx_2.config), Z_CONFIG_LISTEN_KEY, "tcp/127.0.0.1:7449");
+    zp_config_insert(z_loan_mut(node_ctx_2.config), Z_CONFIG_CONNECT_KEY, "tcp/127.0.0.1:7447");
+    zp_config_insert(z_loan_mut(node_ctx_2.config), Z_CONFIG_CONNECT_KEY, "tcp/127.0.0.1:7448");
+    zp_config_insert(z_loan_mut(node_ctx_3.config), Z_CONFIG_MODE_KEY, "peer");
```
[the remark this answers is not in the document]
Comment (jean-roland) [reply tests/z_test_unicast.c:205]: No and it wasn't tested so possibly an edge case that was overlooked.

File: src/collections/list.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (sashacmc) [reply src/collections/list.c:94]: But what's the point of passing two if the previous one obviously determines the next one? If the next one is passed incorrectly there will be a leak or an error. Besides, the calling code must store and evaluate both.

File: include/zenoh-pico/system/common/platform.h (hunk 1 above)
Comment (sashacmc) [reply include/zenoh-pico/system/common/platform.h:84]: You can make a predefinition or put it in a separate header file, passing by void * is a last resort and only if the type can be different, but there we have one type and type cast in each implementation.

File: src/collections/list.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (jean-roland) [reply src/collections/list.c:94]: I'm not sure that would work based on how it's used: https://github.com/jean-roland/zenoh-pico/blob/76e09145f3910d7172ab15ffa28ce3ba19f004fe/src/transport/unicast/read.c#L339

The previous value isn't necessarily valid if we're removing the head & subsequent values. Whereas the current element is always valid since it's the one we're iterating on.

File: include/zenoh-pico/collections/intmap.h [hunk 6]
```diff
@@ -117,11 +118,17 @@ void *_z_int_void_map_iterator_value(const _z_int_void_map_iterator_t *iter);
...
-        return (type *)_z_int_void_map_insert(m, k, v, name##_intmap_entry_elem_free);                          \
+        return (type *)_z_int_void_map_insert(m, k, v, name##_intmap_entry_elem_free, true);                    \
+    }                                                                                                           \
+    static inline type *name##_intmap_insert_push(name##_intmap_t *m, size_t k, type *v) {                      \
+        return (type *)_z_int_void_map_insert(m, k, v, name##_intmap_entry_elem_free, false);                   \
     }                                                                                                           \
     static inline type *name##_intmap_get(const name##_intmap_t *m, size_t k) {                                 \
         return (type *)_z_int_void_map_get(m, k);                                                               \
     }                                                                                                           \
+    static inline _z_list_t *name##_intmap_get_all(const name##_intmap_t *m, size_t k) {                        \
```
[the remark this answers is not in the document]
Comment (jean-roland) [reply include/zenoh-pico/collections/intmap.h:129]: The issue with the int_map iterator is it parses the whole int map while I'm only interested in one of the sub-list. So I'll spend a lot of time looking for the values I'm only interested in.

File: include/zenoh-pico/system/common/platform.h (hunk 1 above)
Comment (jean-roland) [reply include/zenoh-pico/system/common/platform.h:84]: I'm pretty sure I tried everything else and it was a last resort.

File: include/zenoh-pico/collections/intmap.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (sashacmc) [reply include/zenoh-pico/collections/intmap.h:129]: Yes, I understand, my bad, I was confused by the name, but it seems to fit with `intmap_get`.

File: tests/z_test_unicast.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (jean-roland) [reply tests/z_test_unicast.c:205]: Ok so the segfault is because as a client the keyexpr are declared as id but the declare system is broken in the same manner as #909 so fixing it means refactoring this part... which should be done but later so yeah for now it's clear clients & peers should not mix up.

File: include/zenoh-pico/system/common/platform.h (hunk 1 above)
Comment (jean-roland) [reply include/zenoh-pico/system/common/platform.h:84]: mutex rec t are defined in this file only in the case of single thread, in reality they are defined in other files and we end up again in cyclical include hell

Comment (sashacmc) [reply include/zenoh-pico/system/common/platform.h:84]: But as far as I see `_z_socket_wait_event` only used in multithreaded mode (By the way, is it an expected limitation that this mode only works for multithreaded mode? Should we mention about it somewhere?).
So for single-threaded mode it is possible not to define this method at all.

Comment (sashacmc) [reply include/zenoh-pico/system/common/platform.h]: By the way i tried to use _z_mutex_rec_t and don't see any problems.
With _z_transport_unicast_peer_list_t, I agree, it is more difficult. But even if you leave void, at least give it a normal name, not ctx
